### DISCUSSION TITLE
Info Panel enhancements

### DISCUSSION THREAD

**Discussion by baronfel** [ionide/ionide-vscode-fsharp#1531] [Ideas]
In some code I was looking at today, I noticed that the Info Panel had a couple gaps. I thought it might be nice to have a discussion thread where we could centralize discussion about enhancements to the feature.

**Comment by baronfel** [comment]
# Show XML documentation for Functions and Fields on a Record

Currently, the Info Panel doesn't show XML Docs for Functions/Fields/etc of a Record type (and probably other types).

**Comment by MangelMaxime** [comment]
It is funny because I was going to report some issues about it ^^.

InfoPanel doesn't show the comment on type alias.

It seems like it displays the attributes/fields from the aliased type but not the documentation about it.

I think type alias should show the top most documentation.

```fs
module Child =

    /// <summary>
    /// Deepest documentation
    /// </summary>
    type User =
        {
            Name : string
        }

type MyUser = Child.User
```

Should show `Deepest documentation`

```fs
module Child =

    /// <summary>
    /// Deepest documentation
    /// </summary>
    type User =
        {
            Name : string
        }

/// <summary>
/// Top most documentation
/// </summary>
type MyUser = Child.User
```

Should show `Top most documentation`

**Reply by baronfel** [reply]
I don't think that Type Abbreviations (what type aliases are called in the SyntaxTree) can have xml comments:
```fsharp
    | TypeAbbrev of
        detail: ParserDetail *
        rhsType: SynType *
        range: range
```
so your last example can't be done because of the AST.

**Reply by MangelMaxime** [reply]
I trust you on that point, I don't know much of the AST.

I just thought about that because at least on the declaration side Ionide is able to show the tooltip but perhaps the resolution is different as it is in the source file directly.

![image](https://user-images.githubusercontent.com/4760796/116579957-34308580-a913-11eb-832e-abdd3eef1857.png)

**Comment by MangelMaxime** [comment]
When working on FSAC TipFormatter I added lang support for `code` tag.

It make nicer code samples in the tooltips:

```fs
/// <summary>
/// Some documentation
/// </summary>
/// <example>
/// <code lang="fsharp">
/// // Some comment to show coloration
/// let add (x : int) (y : int) = x + y
/// </code>
/// </example>
let myFunc x =  ()
```

![image](https://user-images.githubusercontent.com/4760796/116578495-c5066180-a911-11eb-9022-c283350f8b54.png)

However, the info panel isn't processed the same way by VSCode.

![image](https://user-images.githubusercontent.com/4760796/116578624-e36c5d00-a911-11eb-9814-dacabc267cdf.png)

I am not sure if we can use Onigurama (the textmate grammar parser) from a VSCode extension. If we can, we could re-use [Code-Lightner](https://github.com/MangelMaxime/Code-Lightner) which I use to highlights code snippets in different project.

Or at minima we could use something like [Highlight JS](https://highlightjs.org/)

**Reply by MangelMaxime** [reply]
The highlighting could be applied to the signature section too.

**Comment by MangelMaxime** [comment]
PS: I am throwing ideas but I don't guarantee that they are all good/doable. But if I don't propose them, they will probably never be implemented at least like that they have a chance to be ^^

When writing doc comment it is hard to see how it will looks like.

It could be nice to be able to make the info panel update on content change. So we could visualise the result of the doc comment result "in real time".

Currently we have the following option to control info panel update:

![image](https://user-images.githubusercontent.com/4760796/116580901-1879af00-a914-11eb-8fc5-536e02cc9a78.png)

When setting `onCursorMove` it does update the info panel when editing at certain place.

![info_panel_2](https://user-images.githubusercontent.com/4760796/116582011-3693df00-a915-11eb-9c57-924b89ba9a4b.gif)

I suspect (didn't look at the code) that the info panel is not being updated when editing a doc comment because, we are not under a symbol or keyword and so it is filtered out.

If that's the case, we could detect when we are editing a doc comment and re-ask for the same info panel as the last call. Or for the info panel associated to the type being documented if we can detect it.
